pkg/providers.bzl

```diff
@@ -29,3 +28,69 @@ PackageVariablesInfo = provider(
         "values": "Dict of name/value pairs",
     },
 )
+
+# For the below, attributes always look something like this:
+#
+# ```
+#   {"unix": ["0755", "root", "root"]}
+# ```
+#
+
+# TODO: should all of these be singluar or plural?
+#
+# Seems like there's no harm in specifying a plural, even in rules: it would
+# allow some degree of simplification of rules without needing to specify a
+# custom manifest.
+
+# TODO: should we note exceptions from groups of files/labels?  I feel that
+# information is best encoded at the rule level, which seems to necessitate a
+# "plural" structure as defined below.
+PackageFilesInfo = provider(
+    doc = """Provider representing the installation of one or more files to destination with attributes""",
+    fields = {
+        "attributes": """dict of string -> string list: Attributes to apply to installed file(s)""",
+        "source_dest_map": """dict of File -> string: source-destination map""",
```

Comment:
That's what I thought initially, but then we need the ability to create exclusion lists somewhere outside of `pkg_filegroup` as it is now.  If this is `File`s, then we can do everything in pkg_filegroup [like we're doing now](https://github.com/bazelbuild/rules_pkg/blob/36e8363031c761525a6a917cf7ecb84f6ccb32eb/pkg/experimental/pkg_filegroup.bzl#L457-L465).

If the existing interface is not desirable, what would be the best way to accomplish file exclusion?

I explained some of my reasoning for this [inline](https://github.com/bazelbuild/rules_pkg/pull/246/files/be5badadb629c9c24a5dd772f1f518a7b6c7e8fd#diff-b8ef3be79dde93ffccffae703ab11150f6ab8e887345e165dfc2db91c042981cR39-R47).

WDYT?

Reply:
> Wouldn't the rule producing this provider have the exclusion?

Yes.  The immediate idea is that the producing rule would process the dependencies and produce a source-destination mapping in `PackageFilesInfo` that excludes the excluded labels.  This is more or less clear to me.

Thinking on the `Label` vs `File` question: my understanding is that the proper way to access the provider would be something like this:

```python
for dep in ctx.attr.deps:
    pfi = dep[PackageFilesInfo]
    for src, dest in pfi.source_dest_map:
        # Schedule action for copying src to dest
```

The question is thus: how we figure out where `src` is?

If `src` is a `Label`, it's not immediately clear: the way seems to be to use a [`Target`](https://docs.bazel.build/versions/master/skylark/lib/Target.html) object, however, I couldn't find a way to make a `Target` object from a `Label`.  Also, since `Label`s can ultimately point to more than one location, this further complicates the packaging rules.

If `src` is a `File`, we do not need to do any additional computations in the packaging rules: we just schedule a copy from `src` to `dest` by using the existing path information.

Since we need to get the `File`s anyway, it seems to be easier to just pass them along in the provider and let the packaging rules sort them out than it would be to try have them process `Label`s.  Thus, I chose a map of `File` -> destination here.

Does this make sense?  Am I missing something here?

Reply:
I think we need labels in a depset so that they can be inputs to the pkg_rpm or pkg_tar.
I'm happy to defer the words until we try to implement some of these rules.

Reply:
That's fine with me.  I'll update the documentation string accordingly.